### PR TITLE
allow scram_arch and cmssw version manipulation from the ExternalLHEP…

### PR DESCRIPTION
**Pull request by perrozzi**
…roducer

### CODE REVIEW

**Comment by davidlange6** [comment]
please test

File: GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc [hunk 1]
```diff
@@ -225,7 +225,8 @@ ExternalLHEProducer::beginRunProduce(edm::Run& run, edm::EventSetup const& es)
   
   std::ostringstream eventStream;
   eventStream << nEvents_;
-  args_.push_back(eventStream.str());
```
Comment (bendavid) [inline GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc]: What was actually wrong with using push_back and/or emplace_back?

Comment (perrozzi) [inline GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc]: The string vector args, defined in the job cfg, conventionally only contains the gridpack path (P), so args = [P].
Currently ExternalLHEProducer appends the number of events (N), random seed (S) and number of threads (T) arg,
producing therfore args* = [P,N,S,T] which is expected by run_generic_tarball_cvmfs.sh in this order.

If one adds other parameters (O) to args in the cfg,  args = [P,O] and args* = [P,O,N,S,T] therefore run_generic_tarball_cvmfs.sh breaks.
What I did was to enforce args* = [P,N,S,T,O] in ExternalLHEProducer keeping back-compatibility in run_generic_tarball_cvmfs.sh, further allowing O to be passed (whenever present) to runcmsgrid.sh with
./runcmsgrid.sh $nevt $rnum $ncpu ${@:5}
where ${@:5} means "all the arguments passed to the script starting from the 5th, i.e. O"

I tested and seems to work flawless, I hope is clear.
Let me know if you have further comments or questions

**Comment by perrozzi** [comment]
+1

**Comment by davidlange6** [comment]
hi @perrozzi - this is allowing quite a lot more freedom than suggested by the PR title (basically anything can now be passed). Perhaps its possible to move a bit more of this interface out of the external script and into CMSSW ? [or is that all coming as a second step?]

**Comment by perrozzi** [comment]
Hi David, 
args is defined as a vector of strings so I am just allowing more arguments to be passed to down to runcmsgrid.sh (so far only the gridpack path was used) keeping a defined order for back-compatibility.
Here are a couple of slides to illustrate the complete picture including modification to genproductions (the last slide contains an executive summary)
https://docs.google.com/presentation/d/1A0Cb40zyNgnZ1jppl23WSiHGr1MOiuoz1iscqrKy4hs/edit#slide=id.g2247f7bf2f_0_0

Bottom line is that 
1) by default only one argument will continue to be passed as today, i.e. the gridpack full path, and runcmsgrid.sh will now source the scram_arch and CMSSW version used to create the gridpack

More flexibility will be possible, in case of need, after merging PR in cmssw:

2.1) stick to campaign scram_arch and CMSSW by adding “false” parameter after gridpack full path
2.2) enforce a different combination of scram_arch and cmssw by adding “true” parameter after gridpack full path and then chosen scram_arch and CMSSW

Let me know if anything is not clear or if you have comments.
Luca

**Comment by davidlange6** [comment]
+1

**Pull request merged**